onnxruntime/core/providers/cpu/tensor/scatter_nd.cc

```diff
@@ -0,0 +1,156 @@
+    ScatterND);
+
+template<typename Tind>
+Status ScatterNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) const {
+
+  auto input_tensor  = context->Input<Tensor>(0);
+  auto indice_tensor = context->Input<Tensor>(1);
+  auto update_tensor = context->Input<Tensor>(2);
+  ORT_ENFORCE(input_tensor  != nullptr);
+  ORT_ENFORCE(indice_tensor != nullptr);
+  ORT_ENFORCE(update_tensor != nullptr);
+
+  auto input_shape   = input_tensor->Shape();
+  auto indice_shape  = indice_tensor->Shape();
+  auto update_shape  = update_tensor->Shape();
+  if (indice_shape.NumDimensions() == 0) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+      "indices tensor must has rank larger than 0");
+  }
+
+  auto indice_rank = indice_shape.NumDimensions();
+  auto last_indice_dimension = indice_shape[indice_rank - 1];
+  if (last_indice_dimension > static_cast<int64_t>(input_shape.NumDimensions())) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+      "last dimension of indices must not be larger than rank of input tensor");
+  }
+
+  bool is_update_shape_invalid = [&](){
+    auto update_rank = update_shape.NumDimensions();
+    auto input_rank = input_shape.NumDimensions();
+    if (update_rank < indice_rank - 1) {
+      return true;
+    }
+    if ((update_rank >= indice_rank - 1) &&
+        (indice_rank - 1 >= 0) &&
+        (indice_shape.Slice(0, indice_rank - 1) != update_shape.Slice(0, indice_rank - 1))) {
+      return true;
+    }
+    if ((static_cast<int64_t>(input_rank) > last_indice_dimension) &&
+        (update_rank >= indice_rank - 1) &&
+        (input_shape.Slice(last_indice_dimension) != update_shape.Slice(indice_rank - 1))) {
+      return true;
+    }
+    return false;
+  }();
+  if (is_update_shape_invalid) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+      "updates tensor should have shape equal to indices.shape[:-1] + data.shape[indices.shape[-1]:]. ",
+      "updates shape: ", update_shape, ", indices shape: ", indice_shape, ", data shape: ", input_shape);
+  }
+
+  auto output_tensor = context->Output(0, TensorShape(input_shape));
+
+  const auto* src_base = input_tensor->DataRaw();
+  auto* dst_base = output_tensor->MutableDataRaw();
+  bool is_string_type = input_tensor->DataType() == DataTypeImpl::GetType<std::string>();
+
+  // Re-use input for output. If input/output Tensor* are the same, do not copy.
+  if (src_base != dst_base) {
+    if (is_string_type) {
+      const auto* str_begin = input_tensor->template Data<std::string>();
+      const std::string* str_end = str_begin + input_shape.Size();
+      auto* dst = output_tensor->template MutableData<std::string>();
+      std::copy(str_begin, str_end, dst);
+    } else {
+      memcpy(dst_base, src_base, input_tensor->SizeInBytes());
+    }
+  }
+
+  std::vector<int64_t> element_counts(last_indice_dimension, 0LL); // Number of elements for each input dimension
+
+#ifdef USE_OPENMP
```

Comment:
Do we still use it?

Reply:
I'm not sure... Do you recommend removing it altogether with the below `omp parallel for` as well?

Reply:
it seems still used in some other places https://github.com/microsoft/onnxruntime/blob/02c122d6e478d69ee97b3ecdd0aac5cd4f76ba8c/onnxruntime/contrib_ops/cpu/gather_nd.cc#L48

https://github.com/microsoft/onnxruntime/blob/e769617f75ef2ecd68cc6186998b0f6ba3280597/onnxruntime/core/providers/cpu/nn/pool.cc#L48